examples/virtio/blk_driver_vmm.c

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright 2023, UNSW
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+#include <stddef.h>
+#include <stdint.h>
+#include <microkit.h>
+#include "util/util.h"
+#include "arch/aarch64/vgic/vgic.h"
+#include "arch/aarch64/linux.h"
+#include "arch/aarch64/fault.h"
+#include "guest.h"
+#include "virq.h"
+#include "tcb.h"
+#include "vcpu.h"
+#include "virtio/virtio.h"
+#include "virtio/block.h"
+
+/*
+ * As this is just an example, for simplicity we just make the size of the
+ * guest's "RAM" the same for all platforms. For just booting Linux with a
+ * simple user-space, 0x10000000 bytes (256MB) is plenty.
+ */
+#define GUEST_RAM_SIZE 0x6000000
+
+#if defined(BOARD_qemu_arm_virt)
+#define GUEST_DTB_VADDR 0x47f00000
+#define GUEST_INIT_RAM_DISK_VADDR 0x47000000
+#elif defined(BOARD_odroidc4)
+#define GUEST_DTB_VADDR 0x25f10000
+#define GUEST_INIT_RAM_DISK_VADDR 0x24000000
+#else
+#error Need to define guest kernel image address and DTB address
+#endif
+
+/* Data for the guest's kernel image. */
+extern char _guest_kernel_image[];
+extern char _guest_kernel_image_end[];
+/* Data for the device tree to be passed to the kernel. */
+extern char _guest_dtb_image[];
+extern char _guest_dtb_image_end[];
+/* Data for the initial RAM disk to be passed to the kernel. */
+extern char _guest_initrd_image[];
+extern char _guest_initrd_image_end[];
+/* Microkit will set this variable to the start of the guest RAM memory region. */
+uintptr_t guest_ram_vaddr;
+
+#define MAX_IRQ_CH 64
+int passthrough_irq_map[MAX_IRQ_CH];
+
+static void dummy_ack(size_t vcpu_id, int irq, void *cookie) {
```

Comment:
With https://github.com/au-ts/libvmm/pull/43 this won't be necessary. Pretty much the same functions are now available as part of the `virq_*` API.